cmds-42/usr.etc/nmserver/man/nmserver.8
.TH NMSERVER 8 2/22/88
.CM 4
.SH NAME
nmserver \- Mach network server
.SH SYNOPSIS
.B nmserver
[ -c ] [ -t ] [ -p print_level ]
.SH DESCRIPTION
The 
.I network server
is the program responsible for forwarding Mach IPC over the network.
Currently, it also implements a simple network name service.

.B -c
enables debugging in the
.I cthreads
library.

.B -t
enables procedure call tracing, if the executable was compiled with
this facility turned on.

.B -p
is used to set the level of debugging messages to be printed on the
console.

The network server should be started once and for all at boot time.
It
is shared by all users of the machine.
.SH "SEE ALSO"
nmcontrol(1), netname(3)
.br
"Network Server Design", internal document from the Mach group.
.SH HISTORY
.TP 
22-Feb-88  Daniel Julin (dpj) at Carnegie-Mellon University
Created.


cmds-42/ChangeLog
usr.ucb/from/from.c				mmeyer 8/22/89

Change from to do the from on geteuid(), instead of getlogin() which
looks at who is in /etc/utmp.  

usr.ucb/telnet/telnet.c				mmeyer 8/22/89

Fix for bug #2889.

bin/sed/sed1.c					mmeyer 8/21/89

Fix for bug #2831.

usr.bin/calendar/calendar.{c,sh}		mmeyer 8/21/89

Fix for bug #2794

usr.ucb/Mail					mmeyer 6/27/89

Upgraded to Tahoe sources and fixed a few bugs.

etc/mount/mount.c				mmeyer 6/26/89

Change reversal of mounting of nfs partitions.

etc/init/init.c					mmeyer 6/23/89

Changes for onerror/onoption.

etc/rlogind/rlogind.c				mmeyer 6/14/89

Took out malloc from signal handler.

etc/nu/*					mmeyer 6/13/89

Lee's new nu.

etc/ifconfig/ifconfig.c				mmeyer 6/13/89
etc/bootpd/*

Changes for sexy netinit

bin/awk/awk.g.y					mmeyer 6/1/89

Fixed core dump for Trey.  

etc/quot/quot.c					mmeyer 5/31/89

Changed acct() to print out the correct number of 1024 byte blocks.

bin/sh/homedir.c				mmeyer 5/26/89

Changed sh to use getpwnam() instead of searching through /etc/passwd

usr.ucb/indent/lexi.c				wrp    5/22/89
usr.ucb/indent/indent.c		
usr.ucb/indent/indent_globs.h

Bill Parkhurst's changes for Objective-C 4.0 support, along with
a bugfix that Brian and Morris made.


bin/diff/diffh_vers.c				mike   5/22/89
usr.bin/diff3/Makefile				

Bugfix from Mike DeMoney.  Fixes problem that occurs during integration.

usr.bin/tip/acu.c				mmeyer 5/20/89
usr.bin/tip/cu.c
usr.bin/tip/tip.c
usr.bin/tip/cmds.c

Changed references to connect(), abort() and send() to compile
with shlibs.

bin/csh/sh.c					mmeyer 5/19/89

Integrated the atexit() mod.

usr.ucb/rlogin/rlogin.c				mmeyer 5/19/89

Verified that Mike DeMoney's change to make chars with 0x7f was in.

bin/make/rules.c				mmeyer 5/19/89
bin/make/doname.c	

Integrated Mike's fixes.

usr.etc/mach_swapon/swaptab			mmeyer 5/19/89

Changed to /private/vm/swaptab instead of /etc/vm/swaptab

usr.bin/calendar/Makefile			mmeyer 5/18/89

Changed Makefile so /usr/lib/calendar would install correctly.

etc/route/route.c				mmeyer 5/16/89
etc/rwhod/rwhod.c
usr.etc/nfsstat/nfsstat.c
usr.etc/rpc.rstatd/rpc.rstatd.c
usr.ucb/sysline/sysline.c (not that this will ever run on our system)
usr.ucb/sendmail/src/conf.c
usr.bin/uucp/acucntrl.c

Changed all to nlist "/mach" instead of "/vmunix"

etc/syslog/syslog.c				mmeyer 5/15/89

Changed syslog to print out "mach:" instead of "vmunix:"

bin/date/date.c					mmeyer 5/14/89

Initialized tz.{tz_minuteswest,tz_dsttime} to zero.  This bug
cropped up when linking with shared libraries.  Kevin Enderby
has a bug report.

mk/cmds/usr/ucb/w/Makefile			mmeyer 5/14/89

Linked uptime to w in the INSTALL phase

usr.bin/uucp/fio.c				mmeyer 5/9/89
usr.bin/uucp/gio.c
usr.bin/uucp/logent.c
usr.bin/uucp/tio.c

Ifdef'd syslog() so uucp would link with the libsys_s shlib.
	

usr.bin/uucp/Makefile

Took out duplicate index.c from uucp.

usr.bin/troff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  troff had its own 
version of calloc,realloc,malloc and free.

usr.bin/nroff/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  nroff had its own 
version of calloc,realloc,malloc and free.

usr.ucb/rusers/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  rusers had its own 
version of calloc,realloc,malloc and free.

usr.ucb/msgs/msgs.c	  			mmeyer 5/9/89

Ifdef'd send() so /usr/ucb/msgs would link with the libsys_s shlib.

usr.ucb/more/more.c	  			mmeyer 5/9/89

Ifdef'd strlen() so /usr/ucb/more would link with the libsys_s shlib.

usr.ucb/ex/exrecover.c  			mmeyer 5/9/89

Ifdef'd fprintf() so /usr/ucb/exrecover would link with the libsys_s shlib.

usr.ucb/Mail/aux.c  				mmeyer 5/9/89

Ifdef'd index() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/send.c				mmeyer 5/9/89
usr.ucb/Mail/cmd1.c
usr.ucb/Mail/cmd2.c
usr.ucb/Mail/collect.c
usr.ucb/Mail/edit.c
usr.ucb/Mail/fio.c
usr.ucb/Mail/quit.c

Ifdef'd send() so /usr/ucb/Mail would link with the libsys_s shlib.

usr.ucb/Mail/head.c
usr.ucb/Mail/list.c
usr.ucb/Mail/aux.c

Ifdef'd raise() so /usr/ucb/Mail would link with the libsys_s shlib.

etc/sa/sa.c					mmeyer 5/9/89

Ifdef'd bcmp() so /etc/sa would link with the libsys_s shlib.

etc/routed/defs.h				mmeyer 5/9/89
etc/routed/startup.c
etc/routed/inet.c 

Ifdef'd inet_makeaddr() so /etc/routed would link with the libsys_s shlib.

bin/su/su.c					mmeyer 5/9/89

Unifdef'd getenv() so su would link with the libsys_s shlib.

bin/awk/Makefile				mmeyer 5/9/89

Took out malloc.c and calloc.c from CFILES.  awk had its own 
version of calloc,realloc,malloc and free.

All CMDS Makefiles				mmeyer 5/8/89

Changed all Makefiles to have -O -g for CFLAGS and -lsys_s for
LIBS.  Changes above are for commands that didn't link.

cmds/bin/cp/cp.c				mmeyer 5/8/89
cmds/bin/cat/cat.c

Integrated Avie's mapped file changes.  He claims a 10% performace
win with mapped files.

cmds/bin/make/main.c				mmeyer 5/1/89

Changed main() to have make to a setrlimit(RLIMIT_STACK) to trap
infinitely recursive macros instead of core dumping.

cmds/bin/sh/homedir.c				mmeyer  4/30/89

Changed sh to use getpwent() rather than grep'ing through /etc/passwd.

cmds/bin/tar/tar.c				mmeyer  4/30/89

Changed the "v" flag printf()'s to fprintf(stderr)'s

cmds/etc/icheck/icheck.c			mmeyer  4/30/89

Modified icheck to know about fast symlinks.

cmds/usr.ucb/rlogin/rlogin.c			mmeyer

Added the oob() signal handler for the SIGURG signal.  Data
was getting lost between the socket and the rlogin program.

cmds/bin/csh/sh.c				mmeyer

Added lines 1022 to 1029 in dosource() which check for no operand 
for source flag.

Changed lines 1036 to 1042 in dosource() to just print out an error
using perror() when an open of a source'd file fails instead of not
letting the login session continue.

cmds/bin/passwd/passwd.c			mmeyer

Took out a check against /bin/sh so a user can switch back
and forth from csh to sh and back again.


cmds-42/usr.etc/nmserver/server/nmserver.c
/* 
 * Mach Operating System
 * Copyright (c) 1987 Carnegie-Mellon University
 * All rights reserved.  The CMU software License Agreement specifies
 * the terms and conditions for use and redistribution.
 */
/*
 * nmserver.c
 *
 * $Source: /private/Net/p1/BSD/mmeyer/CMDS/cmds-13/usr.etc/nmserver/server/RCS/nmserver.c,v $
 *
 */

#ifndef	lint
char nmserver_rcsid[] = "$Header: nmserver.c,v 1.2 89/06/07 15:10:53 mmeyer Locked $";
#endif not lint

/*
 * Test program for the network server - implements full IPC service.
 */

/*
 * HISTORY:
 * 21-Jul-88  Daniel Julin (dpj) at Carnegie-Mellon University
 *	NOTIFY: allocate the notify port for the compatibility netmsgserver.
 *
 * 23-Jun-88  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Added printout of a timestamp on startup.
 *
 *  5-Apr-88  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Added code to exec the old netmsgserver if the kernel is not good
 *	enough to support the new netmsgserver. (COMPAT)
 *
 * 29-Mar-88  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Modified logstat initialization so that tracing can be
 *	turned on very early.
 *
 * 14-Mar-88  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Added code to read a configuration file.
 *
 * 27-Feb-88  Daniel Julin (dpj) at Carnegie-Mellon University
 *	Added "-C" switch to control compatibility mode. (COMPAT)
 *
 * 21-Jul-87  Robert Sansom (rds) at Carnegie Mellon University
 *	Added call to cthread_exit() at the end of main.
 *
 * 20-May-87  Robert Sansom (rds) at Carnegie Mellon University
 *	Allow print_level to be set from command line.
 *
 * 16-Apr-87  Daniel Julin (dpj) at Carnegie Mellon University
 *	Fixed to use standard logstat technology for messages.
 *
 *  2-Jan-87  Robert Sansom (rds) at Carnegie Mellon University
 *	Started.
 *
 */

#include <libc.h>
#include <cthreads.h>
#include <stdio.h>
#include <mach.h>
#include <sys/message.h>
#include <sys/signal.h>

#undef	NET_DEBUG
#define	NET_DEBUG	1
#undef	NET_PRINT
#define	NET_PRINT	1

#include "debug.h"
#include "ls_defs.h"
#include "netmsg.h"
#include "network.h"
#include "nm_init.h"

int		cthread_debug;

#if	COMPAT
#define USAGE	"Usage: nmserver/nmthreads [-f config_file] [-c] [-t] [-p #] [-C]"
#else	COMPAT
#define USAGE	"Usage: nmserver/nmthreads [-f config_file] [-c] [-t] [-p #]"
#endif	COMPAT

#if	COMPAT
extern char	compat_server[];
extern int	errno;
extern int	exit_handler();
#endif	COMPAT


/*
 * main
 *
 */
main(argc, argv)
int	argc;
char	**argv;
{
    int		i;
    int		print_level = -1;
#if	COMPAT
    int		compat_mode = -1;
#endif	COMPAT
    char	*config_file = NULL;
    long	clock;

#if	NeXT
#if	PROF
    signal(SIGUSR2, exit);
#endif	PROF
    signal(SIGHUP, SIG_IGN);
#ifdef notdef
    {
	    /*
	     * Make sure that we can drop a core file in case of error
	     */
	     struct rlimit rlim;

	    rlim.rlim_cur = RLIM_INFINITY;
	    rlim.rlim_max = RLIM_INFINITY;
	    setrlimit(RLIMIT_CORE, &rlim);
    }
#endif
#else	NeXT
    /*
     * Initialise the cthreads package.
     */
    cthread_init();
#endif	NeXT

    /*
     * Initialise the stuff needed for debugging.
     */
    if (!(ls_init_1())) {
	panic("ls_init_1 failed.");
    }

    for (i = 1; i < argc; i++) {
	if ((strcmp(argv[i], "-f") == 0) && ((i + 1) < argc)) {
	    i ++;
	    config_file = argv[i];
	}
	else if (strcmp(argv[i], "-t") == 0) tracing_on = 1;
	else if (strcmp(argv[i], "-c") == 0) cthread_debug = 1;
	else if ((strcmp(argv[i], "-p") == 0) && ((i + 1) < argc)) {
	    i ++;
	    print_level = atoi(argv[i]);
	}
#if	COMPAT
	else if (strcmp(argv[i], "-C") == 0) compat_mode = 1;
#endif	COMPAT
	else {
	    fprintf(stderr, "%s\n", USAGE);
	    (void)fflush(stderr);
	    _exit(-1);
	}
    }

    if (ls_read_config_file(config_file) != TRUE) {
	fprintf(stderr,"%s: Invalid configuration file - Exiting\n",argv[0]);
	(void)fflush(stderr);
	_exit(-1);
    }
    if (print_level != -1)
	debug.print_level = print_level;
#if	COMPAT
    if (compat_mode != -1)
	param.compat = compat_mode;
#endif	COMPAT

    /*
     * Initialize syslog if appropriate.
     */
    if (param.syslog)
	openlog("netmsgserver", LOG_PID | LOG_NOWAIT, LOG_USER);

    /*
     * Put out a timestamp when starting
     */
    clock = time(0);
    fprintf(stdout, "Started %s at %24.24s \n", argv[0], ctime(&clock));
    if (param.syslog)
	syslog(LOG_INFO, "Started %s at %24.24s ", argv[0], ctime(&clock));

    fprintf(stdout, "%s: %s %s print_level = %d.\n", argv[0],
			(tracing_on ? "tracing" : ""),
			(cthread_debug ? "cthread_debug" : ""),
			debug.print_level);
    (void)fflush(stdout);

#if	COMPAT
    /*
     * This ugly code to allow the network server to start up on
     * machines that do not have an up-to-date Mach kernel.
     */
    {
	struct thread_basic_info	ti;
	int				count;
	kern_return_t			kr;

	count = sizeof(struct thread_basic_info);
	kr = thread_info(thread_self(), THREAD_BASIC_INFO, &ti, &count);
	if (kr != KERN_SUCCESS) {
		ERROR((msg,"*** This machine does not run a good Mach kernel ***"));
		ERROR((msg,"*** Starting the old netmsgserver standalone ***"));
		execl(compat_server,"old_netmsgserver",0);
		ERROR((msg,"execl of old netmsgserver failed, errno=%d", errno));
		_exit(2);
	}
    }

    /*
     * Fork a process for the old netmsgserver in compatibility mode.
     * We fork here before we are multi-threaded, but the child will wait
     * until we are ready by looking at a flag.
     */
    if (param.compat) {
	extern int	compat_pid;

	/*
	 * First establish signal handlers to kill the child
	 * if we have to exit.
	 */
#define	HANDLER(sig)	{							\
	if (signal(sig,exit_handler) == (int (*)()) -1) {			\
		ERROR((msg,"netname_init.signal(sig) failed, errno=%d"));	\
		panic("netname_init.signal");					\
	}									\
}
	HANDLER(SIGHUP);
	HANDLER(SIGINT);
	HANDLER(SIGQUIT);
	HANDLER(SIGILL);
	HANDLER(SIGIOT);
	HANDLER(SIGEMT);
	HANDLER(SIGFPE);
	HANDLER(SIGBUS);
	HANDLER(SIGSEGV);
	HANDLER(SIGSYS);
	HANDLER(SIGTERM);
	HANDLER(SIGURG);
	HANDLER(SIGIO);
	HANDLER(SIGCHLD);
#undef	HANDLER

	/*
	 * Fork the old server.
	 */
	compat_pid = fork();
	if (compat_pid < 0) {
		ERROR((msg,"netname_init.fork: cannot start an old network server (COMPAT), errno=%d",errno));
		panic("cannot fork");
	}
	if (compat_pid == 0) {
		msg_header_t	sleep_msg;
		kern_return_t	kr;
		int		fd;
		port_t		notify_port;

		/*
		 * Child. Close all non-essential file descriptors,
		 * wait for the go signal,
		 * then exec the old network server.
		 */
		for (fd = getdtablesize(); fd > 2; fd--)
			close(fd);

#if	NOTIFY
		kr = port_allocate(task_self(),&notify_port);
		if (kr != KERN_SUCCESS) {
			ERROR((msg,"[COMPAT,NOTIFY] port_allocate(notify_port) failed, kr=%d",kr));
			_exit(0);
		}
		kr = task_set_special_port(task_self(),TASK_NOTIFY_PORT,notify_port);
		if (kr != KERN_SUCCESS) {
			ERROR((msg,"[COMPAT,NOTIFY] task_set_special_port(notify_port) failed, kr=%d",kr));
			_exit(0);
		}
		(void)port_disable(task_self(),notify_port);
#else	NOTIFY
		notify_port = task_notify();
#endif	NOTIFY

		sleep_msg.msg_size = sizeof(msg_header_t);
		sleep_msg.msg_local_port = notify_port;
		kr = msg_receive(&sleep_msg, RCV_TIMEOUT, 60000);
		if (kr == KERN_SUCCESS) {
			execl(compat_server,"old_netmsgserver",0);
			ERROR((msg,"execl of old netmsgserver failed, errno=%d", errno));
			_exit(0);
		} else {
			ERROR((msg,"old netmsgserver timed-out while waiting for the GO signal, kr=%d\n",kr));
			_exit(0);
		}
	} else {
		/*
		 * Parent.
		 */
		ERROR((msg,"Started old netmsgserver process for compatibility mode, pid=%d",compat_pid));
	}
    }
#endif	COMPAT

    if (nm_init()) {
#if	NeXT
	char buf[BUFSIZ];
	strcpy (buf, "Network Server Initialised.");	
    	DEBUG0(TRUE, 3, buf);
#else	NeXT
	ERROR((msg,"Network Server initialised."));
#endif	NeXT
	DEBUG0(TRUE, 3, 1260);
	DEBUG_NETADDR(TRUE, 3, my_host_id);
    }
    else {
	panic("Network Server initialisation failed.");
    }

#if	PROF
    {
	/*
	 * Wait for ever so that we do not drop off the end of main.
	 */
	port_t		prof_port;
	msg_header_t	prof_msg;
	(void)port_allocate(task_self(), &prof_port);
	_netname_check_in(PORT_NULL, "PROF_EXIT", task_self(), prof_port);
	prof_msg.msg_size = sizeof(prof_msg);
	prof_msg.msg_local_port = prof_port;
	msg_receive(&prof_msg, MSG_OPTION_NONE, 0);
	fprintf(stderr, "Received profile exit message - exitting.\n");
	_exit(1);
    }
#endif	PROF

    cthread_exit(0);

}

